### PR TITLE
Traits

### LINKED ISSUE #7: Implement Trait support

### PR DESCRIPTION
**Pull request by jaredpar**
Adds the ability to filter the test list based on the traits attached to test cases.  

This is my first real venture into WPF based MVVM.  I followed the existing patterns where applicable and consulted tutorials for the new ones I needed to establish (in particular how to communicate selection information).  Let me know if there is anything I got wrong here. 

closes #7

### CODE REVIEW

File: xunit.runner.wpf/MainWindow.xaml [hunk 1]
```diff
@@ -98,7 +98,29 @@
...
+                            </DataTemplate>
+                        </ListBox.ItemTemplate>
+                        <ListBox.ItemContainerStyle>
+                            <Style TargetType="{x:Type ListBoxItem}">
+                                <Setter Property="IsSelected" Value="{Binding Mode=TwoWay, Path=IsSelected}"/>
+                            </Style>
+                        </ListBox.ItemContainerStyle>
+                        <ListBox.ContextMenu>
+                            <ContextMenu>
+                                <MenuItem Header="Clear" Command="{Binding TraitsClearCommand}" />
```
Comment (Pilchie) [inline xunit.runner.wpf/MainWindow.xaml]: The old xunit gui also supported Escape to clear at least some filters.  We should consider that as well as a clunky right-click menu.

File: xunit.runner.wpf/ViewModel/MainViewModel.cs [hunk 2]
```diff
@@ -48,6 +49,8 @@ public MainViewModel()
             this.WindowLoadedCommand = new RelayCommand(OnExecuteWindowLoaded);
             this.RunCommand = new RelayCommand(OnExecuteRun, CanExecuteRun);
             this.CancelCommand = new RelayCommand(OnExecuteCancel, CanExecuteCancel);
+            this.TraitSelectionChangedCommand = new RelayCommand(OnTraitSelectionChanged);
+            this.TraitsClearCommand = new RelayCommand(OnTraitsClear);
```
Comment (Pilchie) [inline xunit.runner.wpf/ViewModel/MainViewModel.cs]: The other command hookups are all named `OnExecuteXXX`

File: xunit.runner.wpf/ViewModel/MainViewModel.cs [hunk 3]
```diff
@@ -362,7 +384,12 @@ private List<ITestSession> RunTests()
         private void OnTestDiscovered(object sender, TestCaseDataEventArgs e)
         {
             var t = e.TestCaseData;
-            allTestCases.Add(new TestCaseViewModel(t.SerializedForm, t.DisplayName, t.AssemblyPath));
+
+            var traitMap = t.TraitMap.Count == 0
```
Comment (Pilchie) [inline xunit.runner.wpf/ViewModel/MainViewModel.cs]: This is no longer a map once you flatten it.

File: xunit.runner.wpf/ViewModel/MainViewModel.cs [hunk 4]
```diff
@@ -402,6 +429,22 @@ private void OnExecuteCancel()
...
+            FilterAfterDelay();
+        }
+
+        private void OnTraitsClear()
+        {
+            foreach (var cur in this.traitCollectionView.Collection)
+            {
+                cur.IsSelected = false;
+            }
+        }
```
Comment (Pilchie) [inline xunit.runner.wpf/ViewModel/MainViewModel.cs]: Do you need to call `FilterAfterDelay` here?

File: xunit.runner.wpf/ViewModel/TraitCollectionView.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        public ObservableCollection<TraitViewModel> Collection => _collection;
+
+        public TraitCollectionView()
+        {
+
+        }
+
+        public void Add(ImmutableArray<TraitViewModel> traitList)
+        {
+            if (traitList.Length == 0)
```
Comment (Pilchie) [inline xunit.runner.wpf/ViewModel/TraitCollectionView.cs]: Is this necessary to protect against getting `default(ImmutableArray<TraitViewModel>)`?  If so, perhaps a more explicit check?

File: xunit.runner.wpf/ViewModel/TraitCollectionView.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+                return;
+            }
+
+            foreach (var traitViewModel in traitList)
+            {
+                TryInsert(traitViewModel);
+            }
+        }
+
+        private void TryInsert(TraitViewModel trait)
```
Comment (Pilchie) [inline xunit.runner.wpf/ViewModel/TraitCollectionView.cs]: There is no point in putting `Try` in the name if you can't observe whether it did anything.  Maybe `InsertIfNotPresent`?

File: xunit.runner.wpf/ViewModel/TraitViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+{
+    public sealed class TraitViewModel : ViewModelBase
+    {
+        private readonly string _name;
+        private readonly string _value;
+        private bool _isSelected;
+
+        public string Name => _name;
+        public string Value => _value;
+        public string DisplayName => $"{Name}={Value}";
```
Comment (Pilchie) [inline xunit.runner.wpf/ViewModel/TraitViewModel.cs]: Just use getter only autoprops?

File: xunit.runner.wpf/ViewModel/TraitViewModel.cs [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+        public bool IsSelected
+        {
+            get { return _isSelected; }
+            set { Set(ref _isSelected, value, nameof(IsSelected)); }
```
Comment (Pilchie) [inline xunit.runner.wpf/ViewModel/TraitViewModel.cs]: The name is `[CallerMemberName]` and so can be omitted.

File: xunit.runner.wpf/ViewModel/TraitViewModelComparer.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace xunit.runner.wpf.ViewModel
+{
+    internal sealed class TraitViewModelComparer : IEqualityComparer<TraitViewModel>, IComparer<TraitViewModel>
+    {
+        internal static readonly TraitViewModelComparer Instance = new TraitViewModelComparer();
+
+        private readonly StringComparer _comparer = StringComparer.Ordinal;
```
Comment (Pilchie) [inline xunit.runner.wpf/ViewModel/TraitViewModelComparer.cs]: Is storing this actually useful?

File: xunit.runner.wpf/ViewModel/MainViewModel.cs (hunk 2 above)
Comment (jaredpar) [inline xunit.runner.wpf/ViewModel/MainViewModel.cs]: Fixed this in a later commit

File: xunit.runner.wpf/ViewModel/MainViewModel.cs (hunk 4 above)
Comment (jaredpar) [inline xunit.runner.wpf/ViewModel/MainViewModel.cs]: Don't believe so.  The clearing of the selection will trigger `OnTraitSelectionChanged` which will call `FilterAfterDelay`.  I tested this at some point last night.  Will verify I didn't later break it.

File: xunit.runner.wpf/ViewModel/MainViewModel.cs (hunk 3 above)
Comment (jaredpar) [inline xunit.runner.wpf/ViewModel/MainViewModel.cs]: Indeed.  Will fix.

File: xunit.runner.wpf/ViewModel/TraitCollectionView.cs (hunk 5 above)
Comment (jaredpar) [inline xunit.runner.wpf/ViewModel/TraitCollectionView.cs]: Yep

File: xunit.runner.wpf/ViewModel/TraitCollectionView.cs (hunk 6 above)
Comment (jaredpar) [inline xunit.runner.wpf/ViewModel/TraitCollectionView.cs]: Good point.

File: xunit.runner.wpf/ViewModel/TraitViewModel.cs (hunk 7 above)
Comment (jaredpar) [inline xunit.runner.wpf/ViewModel/TraitViewModel.cs]: Sure.

**Comment by jaredpar** [comment]
Closing because it's getting too hard to mange this and #22.  Just going to let #22 hold both of the features.  All the feedback will be addressed in that PR.

**Pull request closed without merging**